src/main/java/io/quarkus/security/QuarkusPermission.java

```diff
@@ -0,0 +1,72 @@
+package io.quarkus.security;
+
+import io.quarkus.security.identity.SecurityIdentity;
+
+import java.security.Permission;
+import java.util.Objects;
+
+/**
+ * Special type of the {@link PermissionsAllowed#permission()} that does not require {@link SecurityIdentity}
```

Comment:
@michalvavrik I'm not sure why we are referring to the augmentation not required in this case, since adding custom permission checkers directly in the augmentor is not required, it is optional ?

The other thing that I'm not sure about is whether this class should be here, as opposed to Quarkus.
Adding a custom permission checker to the augmentor is not an API related concept

Reply:
> @michalvavrik I'm not sure why we are referring to the augmentation not required in this case, since adding custom permission checkers directly in the augmentor is not required, it is optional ?

it is hard for me to explain because I thought javadoc was clear and I see I missed there something that makes it hard to understand. maybe, could you look here https://github.com/quarkusio/quarkus/commit/be434b6ef1123237291b8d3baa609c87559f88b1#diff-abf93be6f2bd91b110539ba16ef14044fa71b41e6de14e5f4e6b883de8865a3f there is `QuarkusPermissionSecurityIdentityAugmentor` that is added automatically whenever at least one  `@PermissionsAllowed(permission = MyQuarkusPermission)` is detected where `class MyQuarkusPermission extends QuarkusPermission`. is it clearer?

> The other thing that I'm not sure about is whether this class should be here, as opposed to Quarkus.
Adding a custom permission checker to the augmentor is not an API related concept

but you can say this about any file inside of this project. I think the part you missed (or don't agree with) is that I want to make `QuarkusPermission` an option. Where user can prefer either `@PermissionChecker` or `QuarkusPermission`.

While I suppose user will not be bothered by the fact that `@PermissionChecker` is transformed to the `QuarkusPermission`.

I can totally imagine that user will want to implement `QuarkusPermission` instead of `@PermissionChecker` and I don't see a reason why to take this option away. If we want to be opinionated, we can be opinionated in the docs.

Reply:
>there is QuarkusPermissionSecurityIdentityAugmentor that is added automatically whenever at least one @PermissionsAllowed(permission = MyQuarkusPermission) is detected where class MyQuarkusPermission extends QuarkusPermission. is it clearer?

It is but it was what what I was thinking how it would work. My problem with the current docs wording is that the augmentation in order to add a custom action checker is not required for the permission based control check, users can just add a custom Permission class and the `implies` thing, or have String permission matches done for them. See what I mean ? So this class is about helping users who need to attach custom permission checkers to the identity.




>but you can say this about any file inside of this project.

This project could have been in Quarkus itself but it is not, here it is just a set of interfaces,  with a couple of basic token credential instances. 


>I can totally imagine that user will want to implement QuarkusPermission instead of @PermissionChecker

IMHO we are trying to be super flexible, but it feels like it is a bit early.

IMHO, for a while at least we should keep `QuarkusPermission` a vertx or security extension/runtime level implementation detail. `PermissionChecker` is more than enough to start with, users looking for simplicity will just add this annotation and it works.  

The fact that in the current form, `QuarkusPermission` can only take `SecurityIdentity` in its `isGranted` method makes me think it is too early to start making this class part of the API package... IMHO the whole prospect of the `@PermissionChecker` is not about putting `SecurityIdentity` check to some other place from the augmentor

Reply:
Hey, a lot of text, thank you for taking time to answer. I  believe there are basically 2 questions or proposals that requires answer:

1. Move `QuarkusPermission` to the Quarkus project. It won't be API.

I respect the opinion and let's do it. I have different opinion, but these things doesn't have one right answer.

Nevertheless, whatever `@PermissionChecker` will be able to do, `QuarkusPermission` will be able to do anyway. I don't want to implement new construct when I have one2one relationship that extends current `@PermissionsAllowed` concept.

2. `QuarkusPermission` is just a class that contains `isGranted` and cannot accept method arguments

`QuarkusPermission` is `java.security.Permission`. As every other permission, its constructor can accept every secured method argument it wishes.

I already wrote all the required coverage for QuarkusPermission, I think you can have look here and see all its capabilities:

- https://github.com/quarkusio/quarkus/commit/4d50853fb2e82e6472e45a4f6ea97bb9dbdb6f85
- I recommend to look for the `QuarkusPermissionMethodArgsOnlyTest` that I wrote there

Sidenote: I think it's already an improvement and we should merge to Quarkus separately because it would make `@PermissionChecker` review way easier. But I suppose that is not viable as we would have to release this project before I implement the checker.

Reply:
Hi @michalvavrik,  I guess it would be easier, if this PR would indeed be only about `@PermissionChecker`, and then you can work on a Quarkus branch to support it with `QuarkusPermission`.

Right, so using `QuarkusPermission` on its own is really about making it easy for users to write custom permission checkers, I'm just not sure yet how we should position it yet, can you please type here a simple example, showing the augmentor and secured resource method with 1) the custom Permission class  2) `QuarkusPermission` ? May be it can eventually be represented here as an interface, IMHO it is a separate PR anyway...

Reply:
> Hi @michalvavrik, I guess it would be easier, if this PR would indeed be only about @PermissionChecker, and then you can work on a Quarkus branch to support it with QuarkusPermission.

Yes, I think it would be easier.

> IMHO it is a separate PR anyway...

fine with me, let's not continue here, it's not a big deal, I'll remove it from here

> I'm just not sure yet how we should position it yet, can you please type here a simple example, showing the augmentor and secured resource method with 1) the custom Permission class 2) QuarkusPermission ?

I'm about call it a day. I'll type it tomorrow if you will still want it, but basically only difference is that you need augmentor for a custom permission or you need extra permission that has no args constructor and you map SecurityIdentity roles to this possessed permission. Capibilities are same, you write more code.

Reply:
I just realized that it won't do as I need permission name to be optional. So I'll have to wait for the release with Quarkus PR anyway. Also personally I still think it will be hard to understand that permission name is optional without having `QuarkusPermission` which is only case where it makes sense  as part of the API. Anyway, I am bit tired maybe I'll see tomorrow something else.

Reply:
I suppose I'll have to change `@PermissionsAllowed#value` back to the required as I don't think it makes sense without it. On the Quarkus side I will figure something out, but users won't be able to use `QuarkusPermission`. So only `@PermissionChecker`. np

Reply:
Slept on it. My conclusion if `QuarkusPermission` is removed from here:

- let's make `PermissionsAllowed#value` optional for every non-string `PermissionsAllowed#permission`
  - I showed in the it's capabilities
  - it avoids CDI when you don't need it
  - it is base for the `@PermissionChecker` generated classes
- let's document only `@PermissionChecker`

Reply:
@michalvavrik 

> you need augmentor for a custom permission

OK, I think we'll need to review this idea, I think, in case of the custom permission, it should be enough to augment the identity with the possessed `Permission` only, the check which only does `requiredPermission.impies(possesedPermission)` can be done automatically. I'm not sure why `QuarkusPermission` is needed in this case to avoid it.

As far as the `@PermissionAllowed` and `@PermissionChecker` pair is concerned, so in this case we have this solution: https://github.com/quarkusio/quarkus/commit/43d15c5040ec87bafe0d025379940a461714bafd#diff-abf93be6f2bd91b110539ba16ef14044fa71b41e6de14e5f4e6b883de8865a3f

`QuarkusPermission` is not technically required for this either, we can have a runtime Permission helper which in its implies method calls the PermissionChecker method.

Also, we have to be careful here, here the possessed permission is not even present. I believe we must reject `no-arg` `@PermissionChecker` methods or methods with arguments which do not accept any request parameters or SecurityIdentity, otherwise it will be self-authorization - such checkers can not assert anything about the identity or request.

IMHO,  it is not obvious we need `QuarkusPermission` at all. 
I'm sorry if I keep missing something.

In any case, thanks for starting looking into it, `@PermissionChecker` is a big deal.
I'll have to start preparing for Devoxx Be, I'm a bit late with that, so we'll probably finalize the week starting 14th Oct

Thanks :+1:

Reply:
> requiredPermission.impies(possesedPermission) can be done automatically. I'm not sure why QuarkusPermission is needed in this case to avoid it.

It is needed because it is the other way around: `possesedPermission.impies(requiredPermission)`, it is determined by Java Security API which says that required permission is a subset of possessed permission.

> I'll have to start preparing for Devoxx Be, I'm a bit late with that, so we'll probably finalize the week starting 14th Oct

Pity. Considering Quarkus 3.16 is released by the end of October, I though we could agree on API and release these changes which would allow me to create PR in the Quarkus project and maybe Stephane could review there?

Anyway, Devoxx Be prep is more important. As there won't be anything happening for more than a week, I'll concentrate on other work. Thanks for your time so far.

> Also, we have to be careful here, here the possessed permission is not even present. I believe we must reject no-arg @PermissionChecker methods or methods with arguments which do not accept any request parameters or SecurityIdentity, otherwise it will be self-authorization - such checkers can not assert anything about the identity or request.

Sure, but these are 2 different things - no args subclass of the `java.security.Permission` makes only sense with custom augmentor. While whether accepts `@PermissionChecker` no args can avoided by validation. API doesn't say we are going to support no args checker.

Let's put this PR on hold till you are back.